### ISSUE TITLE
Use of . with HIERARCH keyword

### ISSUE THREAD

**Issue by tony-johnson** [nom-tam-fits/nom-tam-fits#16]
HI, I am trying to use nom.tam.fits to write a card of the form:

HIERARCH MONOCH.WAVELENG= 624.996 / monochromator wavelength

We are doing the following:

FitsFactory.setUseHierarch(true);
hdu.addValue("HIERARCH.MONOCH.WAVELENG",624.996,"monochromator wavelength")

However the card then gets written as

HIERARCH MONOCH WAVELENG=       624.996 / monochromator wavelength

(No . after the MONOCH)

Immediate question:

* Is there any way to write this while preserving the "." in the card.

More philosophical questions:

* Why did nom.tam.fits adopt the convention of putting a . after the HIERARCH in the keyword?
* Wouldn't it be better if the HIEARCH was handled automatcally, i.e.

FitsFactory.setAutoHierarch(true);
hdu.addValue("MONOCH.WAVELENG",624.996,"monochromator wavelength")

would produce

HIERARCH MONOCH.WAVELENG=       624.996 / monochromator wavelength

(because MONOCH.WAVELENG is too long to be a valid keyword).

**Comment by ritchieGitHub** [comment]
thats a question for @TomMcGlynn

**Comment by TomMcGlynn** [comment]
Since the HIERARCH convention is not part of the FITS standard to start with it's a bit dicey to say that this usage is invalid.  However to the extent that this usage <is> documented, e.g., http://fits.gsfc.nasa.gov/registry/hierarch/hierarch.pdf, there is a statement that only characters legal in FITS keywords are allowed in the tokens between HIERARCH and '='. This does not include ".".  There is also an explicit mention that
  HIERARCH A B C = ...
should be thought of as defining an element A.B.C which is why that syntax is used in the library to
specify such keywords.

Is there any documentation of the HIERARCH keyword as you are using it?  If the document referenced above has been replaced, then the library could be updated to match, but changes that violate the documentation seem inappropriate.

**Comment by TomMcGlynn** [comment]
While I'm not really keen on this usage you can probably do something where you create a header card using the current approach, get its string value, insert the '.' and then write the card in the comment style.  E.g. (entirely untested):

    HeaderCard card = new HeaderCard("HIERARCH.MONOCH.WAVELENGTH", 662.3, "My comment");
    String cstr = card.toString().substring(9); // Get rid of HIERARCH
    cstr = cstr.replaceAll("MONOCH WAVELENGTH", "MONOCH.WAVELENGTH");
    header.insertCommentStyle("HIERARCH", cstr);

It's a bit of a kludge but if you're trying to get do something forbidden by the 'standard' then that might be appropriate.

**Comment by tony-johnson** [comment]
The document I was looking at was this one:

http://heasarc.gsfc.nasa.gov/fitsio/c/f_user/node28.html

which states:

"the actual keyword name ('ESO INS FOCU POS' in this example) begins in column 10 and can contain any printable ASCII text characters, including spaces. The equals sign marks the end of the keyword name and is followed by the usual value and comment fields just as in standard FITS keywords."

This seems to contradict the document quoted above, although to be pedantic that document says

"Under the ESO implementation of this convention, each token string that precedes the
equals sign must only contain characters that are legal in formal FITS keywords" 

so does not claim to be a definitive definition of what all implementations should do.

What I am actually trying to do is reproduce the behaviour of an earlier implementation using CFITSIO, so it appears at least that CFITSIO does not prevent a . from being written into the HIERARCH keywords.

**Comment by TomMcGlynn** [comment]
Unfortunately the link they give to the actual definition of the convention does not seem to work -- and I couldn't find it in the wayback machine either.   I'll check with Bill Pence to see if he has a copy. I've not tried to see, but I wouldn't be surprised if our library will read such records but won't write them.  My sense is that that's probably the sweet spot with kludges like what I suggest above when you really need to do something contrary to the nominal convention.

**Comment by ritchieGitHub** [comment]
@tony-johnson  coudl you give us link as @TomMcGlynn requested?

**Comment by tony-johnson** [comment]
I assume you mean the link to the document at:

http://arcdev.hq.eso.org/dicb/dicd/dic-1-1.4.html

referenced by the document I mentioned at:

http://heasarc.gsfc.nasa.gov/fitsio/c/f_user/node28.html?

Unfortunately that link appears to be dead, and I have no way to know what it used to point to.

**Comment by ritchieGitHub** [comment]
The doc in the broken link would be the interesting one, because in all documentation we could find a dot is not a legal character there (as @TomMcGlynn ) said. As @TomMcGlynn  also could not find the document in question, we have first to get the hands on a dokument before implementing this.

For the autohierach request I created a separate issue #19

**Comment by TomMcGlynn** [comment]
I talked with Bill Pence (the author of CFITSIO) about this.  What he indicated was that the ESO convention always precluded the use of the "." as a character within the token.  However when Bill was implementing the convention in CFITSIO he recognized that in terms of what the actual FITS standard allows, any character was permitted in subsequent tokens so that's how he implemented it.  

When writing the Java library I used the ESO convention to define the interface and that explicitly suggests a correspondence between keys of the form A.B.C and cards of the form 
   HIERARCH A B C
so the "." get processed specially.  I don't know how what other implementations of the HIERARCH keywords do.  It might be interesting to see how the original library ESO used (which presumably
predates CFITSIO) handled this, but I don't know if that's available.
